### DISCUSSION TITLE
after_commit event for top level transaction

### DISCUSSION THREAD

**Discussion by andryunin** [sqlalchemy/sqlalchemy#6585] [Usage Questions] [answered]
Hi there!

I'm trying to execute celery task after all DB changes are commited. My idea was to use "after_commit" handler, and it works perfectly fine when not being executed in nested transaction. But in some cases I have to use `begin_nested` blocks, because I process many requests in the loop, and some of them can fail and cause undesirable side effects on models, which without a nested transaction would be commited to database after final `session.commit()` call.

Is there a way to define `after_commit` event handler that fired only on top level commit? Simplified example:
```
session.execute('select 1;')

with session.begin_nested():
    session.execute('select 1;')

    @event.listens_for(session, 'after_commit')
    def after_commit_test(session):
        # Normally here I call celery task
        print('after commit task')

session.commit()
```

Now I get:
```
# BEGIN (implicit) (base.py:711)
# select 1; (base.py:1205)
# SAVEPOINT sa_savepoint_1 (base.py:1205)
# select 1; (base.py:1205)
# RELEASE SAVEPOINT sa_savepoint_1 (base.py:1205)
after commit task
# COMMIT (base.py:753)
after commit task
```

What I want to achieve is to get only one "after commit task" print after final commit.

**Comment by zzzeek** [comment, marked as answer]
you might be better off using the https://docs.sqlalchemy.org/en/14/orm/events.html?highlight=after_transaction_end#sqlalchemy.orm.SessionEvents.after_transaction_end hook, however you can here check the "nested" status in after_commit by looking at ~~"session.get_transaction().nested"~~ `session.in_nested_transaction()` (1.4 and higher).

**Reply by andryunin** [reply]
I tried `after_commit` with `session.transaction.nested` check (still using version 1.3) and it worked. Thank you!

**Reply by AlexanderFarkas** [reply]
That's very strange, but for me session.get_transaction().nested always returns False. I guess that's the reason - nested transaction passes its session which has root transaction context.
<img width="632" alt="Screenshot 2025-02-14 at 10 13 23" src="https://github.com/user-attachments/assets/7f610991-a96a-482f-938c-eef357f4c40f" />

**Reply by AlexanderFarkas** [reply]
Oh, now I get it - session.get_transaction() always returns topmost one. To get the current you need to access private session._transaction. Is it how it supposed to be? @zzzeek

**Reply by CaselIT** [reply]
Can't you use `get_nested_transaction`?

**Reply by zzzeek** [reply]
> Oh, now I get it - session.get_transaction() always returns topmost one. To get the current you need to access private session._transaction. Is it how it supposed to be? @zzzeek

I edited my comment which was incorrect.  use `Session.in_nested_transaction` and `Session.get_nested_transaction`.

**Reply by AlexanderFarkas** [reply]
@zzzeek @CaselIT  Yes, it works on 2.0.38. Thank you very much :)
